core/templates/pages/exploration-player-page/services/stats-reporting.service.ts

```diff
@@ -16,344 +16,416 @@
-    var explorationStarted = false;
-    var explorationActuallyStarted = false;
-    var explorationIsComplete = false;
-    var currentStateName = null;
-    var nextExpId = null;
-    var previousStateName = null;
-    var nextStateName = null;
-
-    var _editorPreviewMode = ContextService.isInExplorationEditorPage();
-    var _questionPlayerMode = ContextService.isInQuestionPlayerMode();
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+
+import { ContextService } from 'services/context.service';
+import { ExplorationPlayerConstants } from
+  'pages/exploration-player-page/exploration-player-page.constants';
+import { MessengerService } from 'services/messenger.service';
+import { PlaythroughService } from 'services/playthrough.service';
+import { SiteAnalyticsService } from 'services/site-analytics.service';
+import { Stopwatch, StopwatchObjectFactory } from
+  'domain/utilities/StopwatchObjectFactory';
+import { UrlInterpolationService } from
+  'domain/utilities/url-interpolation.service';
+
+interface IStateStats {
+  /* eslint-disable camelcase */
+  [stateName: string]: {
+    total_answers_count: number;
+    useful_feedback_count: number;
+    total_hit_count: number;
+    first_hit_count: number;
+    num_times_solution_viewed: number;
+    num_completions: number;
+  };
+  /* eslint-enable camelcase */
+}
+interface IAggregatedStats {
+  /* eslint-disable camelcase */
+  num_starts: number;
+  num_completions: number;
+  num_actual_starts: number;
+  state_stats_mapping: IStateStats;
+  /* eslint-enable camelcase */
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class StatsReportingService {
+  constructor(
+    private http: HttpClient,
+    private contextService: ContextService,
+    private messengerService: MessengerService,
+    private playthroughService: PlaythroughService,
+    private siteAnalyticsService: SiteAnalyticsService,
+    private stopwatchObjectFactory: StopwatchObjectFactory,
+    private urlInterpolationService: UrlInterpolationService) {
+    this._editorPreviewMode = (
+      this.contextService.isInExplorationEditorPage());
+    this._questionPlayerMode = this.contextService.isInQuestionPlayerMode();
+    this.refreshAggregatedStats();
+  }
+
+    explorationId: string = null;
+    explorationTitle: string = null;
+    explorationVersion: number = null;
+    sessionId: string = null;
+    stateStopwatch: Stopwatch = null;
+    optionalCollectionId: string = undefined;
+    statesVisited: Set<string> = new Set();
+    numStatesVisited: number = 0;
+    explorationStarted: boolean = false;
+    explorationActuallyStarted: boolean = false;
+    explorationIsComplete: boolean = false;
+    currentStateName: string = null;
+    nextExpId: string = null;
+    previousStateName: string = null;
+    nextStateName: string = null;
+    _editorPreviewMode: boolean = null;
```

Comment:
Please make these two private and remove leading underscore.

Reply:
Done.